fix: update digest lookup to use versioned RepoTags instead of latest

shivakunv opened the pull request:
Biggest risk in config: 
 1, values.yaml and CSV are managed independently, so renovate can update one without keeping the other in sync.
 2. Same tag, new digest, If the tag  still exists, but upstream re-pushes that same tag to a different image, Renovate may keep the tag the same and only change the digest in csv file 

Current Issue:
Renovate always pulls the digest from the latest tag, which results in a SHA difference.
However, the version in the `values.yaml `file is not updated in this scenario, so the digest in the manifests becomes mismatched and stale.

Solution:
Open a PR only when the RepoTags value in values.yaml changes.
Then search for that exact digest in
`bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml`
instead of using the digest from latest.

check open PRs: 
[https://github.com/shivakunv/gpu-operator/pulls](https://github.com/shivakunv/gpu-operator/pulls)

Commit:
fix: update digest lookup to use versioned RepoTags instead of latest

Signed-off-by: Shiva Kumar (SW-CLOUD) <shivaku@nvidia.com>

tariq1890 commented on bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml:
```diff
@@ -201,37 +201,37 @@ spec:
       image: nvcr.io/nvidia/driver@sha256:35359117c5cdf786694d2fdba2ba038e7f673c5d0243c9ed4dc6cdaf6e675e4a
     - name: device-plugin-image
-      image: nvcr.io/nvidia/k8s-device-plugin@sha256:2d16df5f3f12081b4bd6b317cf697e5c7a195c53cec7e0bab756db02a06b985c
+      image: nvcr.io/nvidia/k8s-device-plugin:v0.18.0@sha256:2d16df5f3f12081b4bd6b317cf697e5c7a195c53cec7e0bab756db02a06b985c
```
why are we adding the version numbers here?

shivakunv replied:
Renovate always pulls the main:latest digest and updates the `sha256`. However, we have a specicfic requirement: 
   _the sha in the `bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml` must match the version specified in `values.yaml`_
   
   To achieve this and to ensure both files are updated to the same version we need an identifier. with this change , renovate should first update the tags/version in `values.yaml `, then use that version to fetch the corresponding `sha256` and update the `bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml` . 
   
   if we dont associate the version with the `sha` renovate will not be able to identify which `sha` belongs to which version and it will use `main:latest `.

rajathagasthya commented on bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml:
```diff
@@ -201,37 +201,37 @@ spec:
     - name: gpu-operator-image
       image: ghcr.io/nvidia/gpu-operator:main-latest
```
What happens to this image when we add the commit SHA as tag in #1898? Will it be ignored?

rajathagasthya reviewed (commented):
This largely looks good to me. Just a couple of questions.

rajathagasthya commented on bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml:
```diff
@@ -201,37 +201,37 @@ spec:
       image: ghcr.io/nvidia/gpu-operator:main-latest
     - name: dcgm-exporter-image
-      image: nvcr.io/nvidia/k8s/dcgm-exporter@sha256:7c0ac4430bb0a5868b7404a0e06c47e02b0375b61aadd614385ad0bc2d43815a
+      image: nvcr.io/nvidia/k8s/dcgm-exporter:4.4.2-4.7.0-distroless@sha256:7c0ac4430bb0a5868b7404a0e06c47e02b0375b61aadd614385ad0bc2d43815a
```
@shivakunv Let's separate changes to this file which introduces a new tagging convention into a different PR. The reason is we may need to check with RedHat to make sure this doesn't break their certification.

Once that change is done, we can run your renovate changes to see if things are updating correctly.

shivakunv commented:
> @shivakunv I took a look at some of the [sample PRs](https://github.com/shivakunv/gpu-operator-shiva/pulls) raised in your fork, and I noticed that only one line gets updated in the `bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml` (the CSV file). Note, that for each container image in the CSV file, there are two lines that need to be updated. For example, for dcgm-exporter, both [this line](https://github.com/NVIDIA/gpu-operator/blob/b7a7b87223bf79afd1881f8d05c6f7bfde94169b/bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml#L204) and [this line](https://github.com/NVIDIA/gpu-operator/blob/b7a7b87223bf79afd1881f8d05c6f7bfde94169b/bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml#L929) need to be updated. Can you confirm that renovate is updating both occurrences? Do you have an example PR that demonstrates this?

@cdesiniotis Thanks!! 
I got carried away by too many changes and accidently removed a required change. 
I have corrected it now and updated the code.

Please check my repo PRs opened by renovate: 
[https://github.com/shivakunv/gpu-operator-shiva/pulls](https://github.com/shivakunv/gpu-operator-shiva/pulls)

rahulait commented on .github/renovate.json:
```diff
@@ -69,37 +73,41 @@
-      ],
-      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-base-ubi9$",
+      "matchPackageNames": ["nvcr.io/nvidia/k8s/dcgm-exporter"],
+      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)-(?<distroMajor>\\d+)\\.(?<distroMinor>\\d+)\\.(?<distroPatch>\\d+)-distroless$",
```
I don't see `distro*` groups in renovate docs: https://docs.renovatebot.com/modules/versioning/regex/

shivakunv replied:
yes. invalid regex string. 
I have corrected it with the proper regex.

rahulait reviewed (approved):
LGTM

rahulait commented on .github/renovate.json:
```diff
@@ -27,36 +27,40 @@
+  "rebaseWhen": "always",
+  "force": {
+    "createPr": true,
+    "createPrEvenIfClosed": true
```
I think thats fine, if its re-using, then this looks fine to me.

shivakunv commented on .github/renovate.json:
```diff
@@ -60,47 +64,67 @@
+    {
+      "description": "Keep only the base gdrcopy version as the OS suffix is added at runtime",
+      "matchPackageNames": ["nvcr.io/nvidia/cloud-native/gdrdrv"],
+      "extractVersion": "^(?<version>v?\\d+\\.\\d+\\.\\d+)(?:-.*)?$",
```
stop version: "v2.5.1"  update to   version: "v2.5.2-ubuntu24.04"

shivakunv commented on .github/renovate.json:
```diff
@@ -60,47 +64,67 @@
+      "versioning": "semver"
+    },
+    {
+      "description": "Skip kubevirt-gpu-device-plugin image update until Renovate preserves the required multi-arch digest",
```
Skip kubevirt-gpu-device-plugin image update until Renovate preserves the required multi-arch digest

check : `fix: use multi-arch digest for kubevirt-dp instead of amd64 digest`


[https://github.com/NVIDIA/gpu-operator/pull/2265](https://github.com/NVIDIA/gpu-operator/pull/2265)

shivakunv commented:
There are three issues that need to be addressed in three different repo:-  ( Its not renovate problem): 

1.` dcgm-exporter `repo:- add `docker push` of the `Docker v2 manifest` alongside the existing `OCI index push`
issue: 
 `docker manifest inspect nvcr.io/nvidia/k8s/dcgm-exporter:4.5.3-4.8.2-distroless `
 observe : `"mediaType": "application/vnd.oci.image.index.v1+json" :` it should be `application/vnd.docker.distribution.manifest.list.v2+json `
 
2. `gdrdrv` repo:- add a version tag` (v2.5.2) `pushed as a `Docker v2 manifest `list covering all OS variants
  `docker manifest inspect nvcr.io/nvidia/cloud-native/gdrdrv:v2.5.2 2`
  Issue: 
     no such manifest: `nvcr.io/nvidia/cloud-native/gdrdrv:v2.5.2`

3. `kubevirt-gpu-device-plugin` repo:-  remove the separate `docker push` of the `amd64-specific manifest`, push only the manifest list

```
curl -sSI \
  -H "Authorization: Bearer ${TOKEN}" \
  -H "Accept: application/vnd.docker.distribution.manifest.v2+json" \
  "https://nvcr.io/v2/nvidia/kubevirt-gpu-device-plugin/manifests/v1.5.0" \
  | grep -iE "http/|content-type|docker-content-digest"
```
```

HTTP/1.1 200 OK
Content-Type: application/vnd.docker.distribution.manifest.v2+json
docker-content-digest: sha256:98ff0ae43263b8ab4255197462f0b5d571eed9f263ec5f611f4b98e6fb2920d3
x-content-type-options: nosniff
```
 
Issue: `98ff0ae43263b8ab4255197462f0b5d571eed9f263ec5f611f4b98e6fb2920d3` : should be manifest list not amd64 

@rajathagasthya : PR is ready to merge

rajathagasthya commented:
@shivakunv Can you create issues in the respective repos for tracking?

shivakunv commented:
> @shivakunv Can you create issues in the respective repos for tracking?

created : 
[https://github.com/NVIDIA/dcgm-exporter/issues/676](https://github.com/NVIDIA/dcgm-exporter/issues/676)
[https://github.com/NVIDIA/gdrcopy/issues/340](https://github.com/NVIDIA/gdrcopy/issues/340)
[https://github.com/NVIDIA/kubevirt-gpu-device-plugin/issues/188](https://github.com/NVIDIA/kubevirt-gpu-device-plugin/issues/188)

rajathagasthya commented on deployments/gpu-operator/values.yaml:
```diff
@@ -286,7 +286,7 @@ dcgm:
   repository: nvcr.io/nvidia/cloud-native
   image: dcgm
-  version: 4.5.2-1-ubuntu22.04
+  version: 4.5.2-1-ubi9
```
Didn't we decide not to make this change? Correct me if I'm wrong please.

shivakunv replied:
If you'd like, I can remove this change from `values.yaml,` but if we want to make an OS independent change like the other packages, we'll need a bare tag for `DCGM`.

issue is similar to : 
https://github.com/NVIDIA/gdrcopy/issues/340 .

shivakunv replied:
reverted

rahulait commented:
@shivakunv can we rebase it with latest main? There are a bunch of changes done recently, so some of the changes might not be needed anymore.

shivakunv commented:
> @shivakunv can we rebase it with latest main? There are a bunch of changes done recently, so some of the changes might not be needed anymore.

@rahulait : PTAL

rahulait commented on .github/renovate.json:
```diff
@@ -27,31 +27,35 @@
+  "force": {
+    "createPr": true,
+    "createPrEvenIfClosed": true
+  },
```
I see following in renovate logs:
```
[14:39:23.526] INFO (8): Initialize tools all succeded in 159ms.
 INFO: Renovate started
       "renovateVersion": "43.278.2"
 WARN: Config validation warnings found
       "configType": "/github-action/renovate.json",
       "warnings": [
         {
           "topic": "Configuration Error",
           "message": "Invalid configuration option: createPr"
         },
         {
           "topic": "Configuration Error",
           "message": "Invalid configuration option: createPrEvenIfClosed"
         }
       ]
```

shivakunv replied:
Thanks . removed

rahulait commented on .github/renovate.json:
```diff
@@ -132,18 +136,26 @@
+    {
+      "description": "Keep only the base gdrcopy version as the OS suffix is added at runtime",
+      "matchPackageNames": ["nvcr.io/nvidia/cloud-native/gdrdrv"],
+      "extractVersion": "^(?<version>v?\\d+\\.\\d+\\.\\d+)(?:-.*)?$",
```
Current rules are skipping gdrdrv

```
Dependency nvcr.io/nvidia/cloud-native/gdrdrv has unsupported/unversioned value v2.6 (versioning=semver)
Skipping ... gdrdrv because no currentDigest or pinDigests
```

The new rule does not support its current two-component version (v2.6); it expects vX.Y.Z

https://catalog.ngc.nvidia.com/orgs/nvidia/cloud-native/containers/gdrdrv/-

shivakunv replied:
issue created for gdrdrv: 
[https://github.com/NVIDIA/gdrcopy/issues/340](https://github.com/NVIDIA/gdrcopy/issues/340 )

till than I disabled : sha update: 
```
 "matchFileNames": ["bundle/manifests/gpu-operator-certified.clusterserviceversion.yaml"],
      "matchPackageNames": ["nvcr.io/nvidia/cloud-native/gdrdrv"],
      "enabled": false 
```

rahulait replied:
Sure. Even if it gets fixed, it will still ignore it as the version looks for X.Y.Z and the latest version is X.Y only. Please make sure to keep track of it and fix it when its resolved.

rahulait replied:
Ah nvm, I see you also updated the logic. Sounds good then.

rahulait commented:
@shivakunv  Renovate’s dry run now finds one unrelated available update: nvidia-fs from `2.27.3` to `2.28.2-ubuntu24.04` which is released today.

Similar to gdrcopy, do we need same ticket/issue/change for gds as well?

shivakunv commented:
> @shivakunv Renovate’s dry run now finds one unrelated available update: nvidia-fs from `2.27.3` to `2.28.2-ubuntu24.04` which is released today.
> 
> Similar to gdrcopy, do we need same ticket/issue/change for gds as well?

same issue, bare tag not present
